dev/test/serializer.ts

```diff
@@ -0,0 +1,249 @@
+            ],
+          },
+        },
+      ],
+    };
+
+    validateUserInput('nestedObject', nestedObject, 'Firestore Object', {
+      allowDeletes: 'none',
+      allowTransforms: false,
+      allowUndefined: false,
+    });
+  });
+
+  it('validates the depth of nested objects and arrays - 21', () => {
+    // This nested object is 21 levels deep
+    const nestedObject = {
+      // depth 0
+      links: [
+        // depth 1
+        {
+          // depth 2
+          child: {
+            // depth 3
+            links: [
+              // depth 4
+              {
+                // depth 5
+                child: {
+                  // depth 6
+                  links: [
+                    // depth 7
+                    {
+                      // depth 8
+                      child: {
+                        // depth 9
+                        links: [
+                          // depth 10
+                          {
+                            // depth 11
+                            child: {
+                              // depth 12
+                              links: [
+                                // depth 13
+                                {
+                                  // depth 14
+                                  child: {
+                                    // depth 15
+                                    links: [
+                                      // depth 16
+                                      {
+                                        // depth 17
+                                        child: {
+                                          // depth 18
+                                          uiData: {
+                                            // depth 19
+                                            choicesFactors: {
+                                              // depth 20
+                                              rarely: {
+                                                cat: true, // depth 21
+                                              },
+                                            },
+                                          },
+                                        },
+                                      },
+                                    ],
+                                  },
+                                },
+                              ],
+                            },
+                          },
+                        ],
+                      },
+                    },
+                  ],
+                },
+              },
+            ],
+          },
+        },
+      ],
+    };
+
+    expect(() =>
+      validateUserInput('nestedObject', nestedObject, 'Firestore Object', {
+        allowDeletes: 'none',
+        allowTransforms: false,
+        allowUndefined: false,
+      })
+    ).to.throw(
+      'Value for argument "nestedObject" is not a valid Firestore Object. Input object is deeper than 20 levels or contains a cycle.'
```

Comment:
Consider using a regular expression here instead of an exact match on the exception's message (also below).

go/breaking-tests#appendix-common-types-of-brittleness-and-suggested-resolutions

Reply:
Updated these to regular expressions that will be a little more robust. Thanks!